### DISCUSSION TITLE
Downgrade version

### DISCUSSION THREAD

**Discussion by Arctomachine** [vercel/next.js#38489] [Help] [answered]
How can I downgrade version from 12.1.6 to 12.1.0? I want to use it because the bug introduced in later versions is not fixed yet.
If I just install it over current version, build script gives error
```
> Build error occurred
TypeError: Cannot set properties of undefined (setting 'reactRoot')
```

**Comment by pzi** [comment]
`npm install next@12.1.0` / `yarn install next@12.1.0`

**Reply by Arctomachine** [reply]
I think you misunderstood me. It does not work.

**Comment by icyJoseph** [comment, marked as answer]
You also need to downgrade React from 18 to 17, that `createRoot` API is from React 18.

Next.js is a React framework, so those two go hand in hand. Recent versions of Next.js detect React 18, and switch to the new rendering root automatically.

With yarn:

```
yarn add next@12.1.0 react@17.0.2 react-dom@17.0.2
```

or with npm

```
npm install next@12.1.0 react@17.0.2 react-dom@17.0.2
```

Just to be sure I'd also delete the `.next` and `node_modules`. If you are using TypeScript, you need to downgrade the React types as well.

What bug are you talking about though?

**Reply by Arctomachine** [reply]
Doing this fixed the issue, but it brings another error in `_app.tsx`:
```
TS2786: 'Component' cannot be used as a JSX component.
Its element type 'ReactElement<any, any> | Component<{}, any, any> | null' is not a valid JSX element.
Type 'Component<{}, any, any>' is not assignable to type 'Element | ElementClass | null'.
...
```
```json
    "next": "12.1.0",
    "react": "17.0.2",
    "react-dom": "17.0.2",


    "@types/react": "17.0.2",
    "@types/react-dom": "17.0.2",
    "eslint-config-next": "12.1.0",
```
This is old problem and it was reported, but not resolved as it was TS compatibility problem.

The bug I mentioned is https://github.com/vercel/next.js/issues/38232

**Reply by icyJoseph** [reply]
Ah right that TypeScript bug... You can fix it, temporarily, with:

```json
"resolutions": {
    "**/@types/react": "^17.0.21",
    "**/@types/react-dom": "^17.0.9"
}
```

In your `package.json` - once they fix the other bug, you can stay on React 17 if you like, or upgrade altogether, don't forget to remove the resolutions though.

**Reply by Arctomachine** [reply]
Now it all works. Except I am using pnpm and had to remove `**/` part from resolutions names for it to work. Thanks.

**Comment by najathi** [comment]
With npm:
```
npm i next@12.3.0 react@17.0.2 react-dom@17.0.2 
npm i --save-dev  @types/react@17.0.21 @types/react-dom@17.0.9

```
With yarn:
```
yarn add next@12.3.0 react@17.0.2 react-dom@17.0.2 
yarn add -D @types/react@17.0.21 @types/react-dom@17.0.9
```
